src/poseidon/matrix.rs

```diff
@@ -0,0 +1,675 @@
+//! acknowledgement: adapted from FileCoin Project: https://github.com/filecoin-project/neptune/blob/master/src/matrix.rs
+use std::iter::FromIterator;
+use std::ops::{Index, IndexMut};
+
+use ark_ff::PrimeField;
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct Matrix<T: Clone>(pub Vec<Vec<T>>);
```

Comment:
I would suggest using a `Vec<T>` in row-major order for better performance.

Reply:
make sense (for better cache performance)